File: app/javascript/shared/tables/citations.js
import escape from "lodash/escape";
import {text_renderer, date_time_renderer, earnings_renderer, yes_no_renderer, int_renderer, see_more_renderer, url_renderer, unknown} from "renderers";
import {initDataTable} from "shared/tables/common";

initDataTable("table.citations-table", function($table, options) {
  $table.DataTable({
    ...options,
    columns: [
      {data: "ticket_number", render: url_renderer(text_renderer)},
      {data: "links", render: yes_no_renderer, orderable: false},
      {data: "event_date", render: date_time_renderer},
      {data: "officer_number", render: officer_renderer, orderable: false},
      {data: "issuing_agency", render: agency_renderer},
      {data: "location_name", render: text_renderer},
      {data: "amount", render: earnings_renderer},
      {data: "ticket_type", render: text_renderer},
      {data: "source", render: text_renderer},
      {data: "violator_type", render: text_renderer},
      {data: "cdl", render: yes_no_renderer},
      {data: "license_class", render: text_renderer},
      {data: "posted_speed", render: int_renderer},
      {data: "violation_speed", render: int_renderer},
      {data: "posted", render: yes_no_renderer},
      {data: "radar", render: text_renderer},
      {data: "clocked", render: text_renderer},
      {data: "race", render: text_renderer},
      {data: "sex", render: text_renderer},
      {data: "make", render: text_renderer},
      {data: "model_year", render: int_renderer},
      {data: "vehicle_color", render: text_renderer},
      {data: "sixteen_pass", render: yes_no_renderer},
      {data: "haz_mat", render: yes_no_renderer},
      {data: "paid", render: yes_no_renderer},
      {data: "hearing_requested", render: yes_no_renderer},
      {data: "court_name", render: text_renderer, orderable: false},
      {data: "age", render: int_renderer},
      {data: "searched", render: yes_no_renderer},
      {data: "offenses", render: see_more_renderer, orderable: false},
    ],
    order: [[2, 'desc']]
  });
});

export function agency_renderer(data, type, row) {
  if (type != "display") {
    return data;
  }
  if (data == null) {
    return unknown();
  }
  return `<span>${escape(data.replace(/^Boston Police /, ''))}</span>`;
}

export function officer_renderer(data, type, row) {
  if (type != "display") {
    return data;
  }
  if (data == null) {
    return unknown();
  }
  if (row.officer_name && row.officer_url) {
    return escape(row.officer_name);
    // return `<a href="${escape(row.officer_url)}">${escape(row.officer_name)} (${escape(data)})</a>`;
  }
  return escape(data);
}

File: app/lib/exporter/citations.rb
class Exporter::Citations < Exporter::Exporter
  def column_definitions
    column("ticket_number") { record.citation.ticket_number }
    column("issuing_agency") { record.citation.issuing_agency }
    column("officer_number") { record.citation.officer_number }
    column("ticket_type") { record.citation.ticket_type }
    column("source") { record.citation.source }
    column("violator_type") { record.citation.violator_type }
    column("cdl") { record.citation.cdl }
    column("license_class") { record.citation.license_class }
    column("event_date") { record.citation.event_date }
    column("location_id") { record.citation.location_id }
    column("location_name") { record.citation.location_name }
    column("posted_speed") { record.citation.posted_speed }
    column("violation_speed") { record.citation.violation_speed }
    column("posted") { write_boolean(record.citation.posted) }
    column("radar") { write_boolean(record.citation.radar) }
    column("clocked") { write_boolean(record.citation.clocked) }
    column("race") { record.citation.race }
    column("sex") { record.citation.sex }
    column("vehicle_color") { record.citation.vehicle_color }
    column("make") { record.citation.make }
    column("model_year") { record.citation.model_year }
    column("sixteen_pass") { write_boolean(record.citation.sixteen_pass) }
    column("haz_mat") { write_boolean(record.citation.haz_mat) }
    column("amount") { write_money(record.citation.amount) }
    column("paid") { write_boolean(record.citation.paid) }
    column("hearing_requested") { write_boolean(record.citation.hearing_requested) }
    column("court_code") { record.citation.court_code }
    column("age") { record.citation.age }
    column("searched") { write_boolean(record.citation.searched) }
    column("offense") { record.offense.offense }
    column("description") { record.offense.description }
    column("assessment") { write_money(record.offense.assessment) }
    column("expected_assessment") { write_money(record.offense.expected_assessment) }
    column("display_assessment") { write_money(record.offense.display_assessment) }
    column("disposition") { record.offense.disposition }
    column("disposition_description") { record.offense.disposition_description }
    column("major_incident") { write_boolean(record.offense.major_incident) }
    column("surchargeable") { write_boolean(record.offense.surchargeable) }
    column("sdip_points") { record.offense.sdip_points }
    prefix("officer", Exporter::Officers) { record.citation.officer }
  end

  def records
    Citation.includes(:officer, officer: [:zip_code, :complaint_officers, :complaints, complaint_officers: [:complaint]]).flat_map do |cit|
      cit.offenses.map do |off|
        OpenStruct.new offense: off, citation: cit
      end
    end
  end
end
